homeassistant/components/elevenlabs/tts.py

```diff
@@ -0,0 +1,117 @@
+"""Support for the ElevenLabs text-to-speech service."""
+
+from __future__ import annotations
+
+from functools import cached_property
+import logging
+from typing import Any
+
+from elevenlabs.client import AsyncElevenLabs
+from elevenlabs.core import ApiError
+from elevenlabs.types import Model, Voice
+
+from homeassistant.components import tts
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import CONF_MODEL, CONF_VOICE, DEFAULT_MODEL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def get_model_by_id(client: AsyncElevenLabs, model_id: str) -> Model | None:
+    """Get ElevenLabs model from their API by the model_id."""
+    models = await client.models.get_all()
+    for maybe_model in models:
+        if maybe_model.model_id == model_id:
+            return maybe_model
+    return None
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up ElevenLabs tts platform via config entry."""
+    client = config_entry.runtime_data.client
+    # Get model, fallback to default
+    model_id = config_entry.options.get(CONF_MODEL, DEFAULT_MODEL)
+    model = await get_model_by_id(client, model_id)
+    assert model is not None, "Model was not found in async_setup_entry"
+    voices = (await client.voices.get_all()).voices
+    default_voice_id = config_entry.options.get(CONF_VOICE, voices[0].voice_id)
+    async_add_entities(
+        [ElevenLabsTTSEntity(config_entry, client, model, voices, default_voice_id)]
+    )
+
+
+class ElevenLabsTTSEntity(tts.TextToSpeechEntity):
+    """The ElevenLabs API entity."""
+
+    def __init__(
+        self,
+        config_entry: ConfigEntry,
+        client: AsyncElevenLabs,
+        model: Model,
+        voices: list[Voice],
+        default_voice_id: str,
+    ) -> None:
+        """Init ElevenLabs TTS service."""
+        self._client = client
+        self._model = model
+        self._default_voice_id = default_voice_id
+        self._voices = sorted(
+            (tts.Voice(v.voice_id, v.name) for v in voices if v.name),
+            key=lambda v: v.name,
+        )
+        # Default voice first
+        voice_indices = [
+            idx for idx, v in enumerate(self._voices) if v.voice_id == default_voice_id
+        ]
+        if voice_indices:
+            self._voices.insert(0, self._voices.pop(voice_indices[0]))
+        self._attr_name = config_entry.title
```

Comment:
Please use `device_info` to create a device with entry_type -> Servce, and then leave the name empty. All new integrations should use `_attr_has_entity_name = True`. Then also add `_attr_name = None` so the entity takes the name of the device, which will take the name of the config entry

Reply:
Please, could you tell me why we need this separate Service instead of the default tts.speak one? I am most likely missing something here. This code is heavily derived from the google translate tts integration, so maybe part of the things I did are not up to date

Reply:
Oh, its not a service as in a service call, rather a device that is a service. So if you open your integration list on your environment, you see a lot of x DEVICES and x SERVICES underneath each integration, this makes it a service